### PR TITLE
feat(haystack): add instrumentation support

### PR DESCRIPTION
**Pull request by 21ShisodeParth**
[giant PR]
a first pass at a Haystack instrumentor; the README file will be properly written shortly, just copied from guardrails rn

### CODE REVIEW

**Comment by axiomofjoy** [comment]
Hey @21ShisodeParth, can you sign the CLA and rename your PR in keeping with [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/)? "feat: " is the correct prefix.

**Comment by axiomofjoy** [comment]
Let me know if you need help to pass CI.

File: python/instrumentation/openinference-instrumentation-haystack/CHANGELOG.md [hunk 1]
```diff
@@ -0,0 +1,10 @@
+# Changelog
+
+## 0.1.0 (2024-07-15)
+
+
+### Features
+
+* **guardrailsai:** add instrumentation support ([#553](https://github.com/Arize-ai/openinference/issues/553)) ([fb07e84](https://github.com/Arize-ai/openinference/commit/fb07e84b2027f06ad1c0659b5eabee87b23a0de0))
```
Comment (harrisonchu) [inline python/instrumentation/openinference-instrumentation-haystack/CHANGELOG.md:8]: copy pasta

File: python/instrumentation/openinference-instrumentation-haystack/README.md [hunk 2]
```diff
@@ -0,0 +1,75 @@
+# OpenInference Haystack Instrumentation
+
+[![pypi](https://badge.fury.io/py/openinference-instrumentation-guardrails.svg)](https://pypi.org/project/openinference-instrumentation-guardrails/)
+
+Python auto-instrumentation library for LLM applications implemented with Guardrails
```
Comment (harrisonchu) [inline python/instrumentation/openinference-instrumentation-haystack/README.md:5]: copy pasta

Comment (harrisonchu) [reply python/instrumentation/openinference-instrumentation-haystack/README.md:5]: don't forget to update the example with a haystack specific one.

Comment (harrisonchu) [reply python/instrumentation/openinference-instrumentation-haystack/README.md:5]: for the readme here just choose the simplest possible example

File: python/instrumentation/openinference-instrumentation-haystack/pyproject.toml [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    "openinference-instrumentation>=0.1.7",
+    "openinference-semantic-conventions",
+    "wrapt",
+    "typing-extensions",
+]
+
+
+[project.optional-dependencies]
+instruments = [
+    #"guardrails-ai >= 0.4.5",
```
Comment (harrisonchu) [inline python/instrumentation/openinference-instrumentation-haystack/pyproject.toml:39]: remove comment and add version. you can put >= the version of haystack you implemented on

File: python/instrumentation/openinference-instrumentation-haystack/src/openinference/instrumentation/haystack/_wrappers.py [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+        ) as span:
+            try:
+                response = wrapped(*args, **kwargs)
+            except Exception as exception:
+                span.set_status(trace_api.Status(trace_api.StatusCode.ERROR, str(exception)))
+                span.record_exception(exception)
+                raise
+            span.set_status(trace_api.StatusCode.OK)
+
+            attributes = dict(
```
Comment (harrisonchu) [inline python/instrumentation/openinference-instrumentation-haystack/src/openinference/instrumentation/haystack/_wrappers.py:83]: i would put the setting the `invocation_parameters` attributes above the `response = wrapped(*args, **kwargs)` 

consider the fact that if that line returns an exception, none of the invocation parameters would be set, which would be super debugging information exactly when the exception is happening!

File: python/instrumentation/openinference-instrumentation-haystack/src/openinference/instrumentation/haystack/_wrappers.py [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+            span.set_attributes(dict(get_attributes_from_context()))
+            try:
+                response = wrapped(*args, **kwargs)
+            except Exception as exception:
+                span.set_status(trace_api.Status(trace_api.StatusCode.ERROR, str(exception)))
+                span.record_exception(exception)
+                raise
+            span.set_status(trace_api.StatusCode.OK)
+
+            attributes = dict(
```
Comment (harrisonchu) [inline python/instrumentation/openinference-instrumentation-haystack/src/openinference/instrumentation/haystack/_wrappers.py:196]: same comment re setting the invocation parameters before the wrapped function is executed

File: python/instrumentation/openinference-instrumentation-haystack/src/openinference/instrumentation/haystack/__init__.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+            wrapper = _PipelineWrapper(tracer=tracer),
+        )
+        # Creating child spans for every Component in the Pipeline
+        wrap_function_wrapper(
+            module = "haystack.core.pipeline.pipeline",
+            name = "Pipeline._run_component",
+            wrapper = _ComponentWrapper(tracer=tracer),
+        )
+
+    def _uninstrument(self, **kwargs):
```
Comment (harrisonchu) [inline python/instrumentation/openinference-instrumentation-haystack/src/openinference/instrumentation/haystack/__init__.py:47]: missing uninstruments here

File: python/instrumentation/openinference-instrumentation-haystack/src/openinference/instrumentation/haystack/utils.py [hunk 7]
```diff
@@ -0,0 +1,25 @@
+'''
```
Comment (harrisonchu) [inline python/instrumentation/openinference-instrumentation-haystack/src/openinference/instrumentation/haystack/utils.py:1]: is this still being used?

File: python/instrumentation/openinference-instrumentation-haystack/src/openinference/instrumentation/haystack/__init__.py (hunk 6 above)
Comment (harrisonchu) [reply python/instrumentation/openinference-instrumentation-haystack/src/openinference/instrumentation/haystack/__init__.py:47]: you also want to add logic to check that these functions were indeed wrapped by us before uninstrumenting

check out this example
https://github.com/Arize-ai/openinference/pull/622/files#diff-743fb0121267b7d5919416f1372ebc7d063a9c89cc2b7ae64ccceb3df519c9a6R61-R75

**Review by harrisonchu — COMMENTED** [review]
some commets

**Comment by 21ShisodeParth** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by 21ShisodeParth** [comment]
recheck

**Comment by 21ShisodeParth** [comment]
recheck

**Comment by 21ShisodeParth** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by 21ShisodeParth** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by 21ShisodeParth** [comment]
recheck

**Comment by Parth** [comment]
Fix your bot, it's tagging the wrong people

why do you even feel the need to even have a CLA, probably just a waste of everyone's time

**Comment by axiomofjoy** [comment]
Closing in favor of #644 @21ShisodeParth

**Pull request closed without merging**